accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,632 @@
+
+### Release attestations
+When a package is successfully published, the npm registry will "attest" the release and upload this to Rekor. This includes a signature that links the repository to the package name. This must be verified by the CLI when during verification.
+
+This can be achieved by publishing the package signature to Rekor by using the proposed [“package entry”](https://github.com/sigstore/rekor/pull/805) log type.
+
+The release attestation will contain the following claims, linking the build to the release:
+- Package org/name
+- Package version
+- Registry signature
+- Tarball shasum
+
+### Logging additional events on the Rekor transparency log
+
+The current proposal only suggests publishing two events to the public Rekor log: sign during build and release/publish.
+
+Publishing a richer set of events to the log could help create a safe and independent history of sensitive events. This could provide the possibility of detection when the registry DB or blob bucket are compromised, because it would be out of sync with the transparency log.
+
+There's interest from OpenSSF's Securing Software Repos WG to standardize on a set of events to easy interoperability between different package registries.
+
+Possible events that could be recorded in the transparency log:
+
+- User registration
+- Sign during build
+- Push
+- Publish or release
+- Yank or delete
+- Owner added
+- Owner removed
+- Owner changed email address
+- Owner enabled MFA
+- Owner disabled MFA
+- Endorsement by non-owner
+- Ownership call
+- Ownership request
+- Ownership transfer
+- Resurrection of name
+
+## Prior Art
+### Sigstore cosign CLI
+The [cosign](https://github.com/sigstore/cosign) CLI tool supports signing arbitrary blobs and could be used to sign the package tarball manually but requires significant setup from maintainers and offers no simple way for consumers to verify signatures unless you know you to query for the package signature on Rekor.
+
+### SLSA Provenance Generator
+Google is working on a language agnostic reusable Actions workflow that can generate build provenance claims using Sigstore: [slsa-github-generator](https://github.com/slsa-framework/slsa-github-generator)
+
+It's technically possible to use something like this to produce verifiable build provenance when publishing an npm package but there's similarly to above, no easy way for package consumers to verify signatures in a standardized way when installing.
+
+### SLSA Provenance Verifier
+
+Implementation for verifying [SLSA Provenance](/Users/feelepxyz/src/github.com/npm/rfcs/accepted/0000-link-packages-to-source-and-build.md). It currently supports verifying provenance generated by the [SLSA generator for Go projects](https://github.com/slsa-framework/slsa-github-generator/blob/main/.github/workflows/builder_go_slsa3.yml).
+
+## Unanswered questions
+
+- How could we support legacy packages that might not be maintained anymore?
```

Comment:
Is rebuilding legacy packages off the table for NPM? I know that takes the original maintainer out of the loop. I just stat that because OpenSSF's Alpha Omega is open to potentially funding the safe building of packages.

Reply:
I believe npm packages are considered immutable once published, and that the npm packages are not always reproducible.

A question: are there any circumstances where the source code might not be available? e.g. perhaps a binary could be embedded in an npm package without sharing the source?

Reply:
Tons. Including packages whose repos are no longer available, or didn't contain a repo link at all.

Reply:
> Is rebuilding legacy packages off the table for NPM?

No but we won't be doing it as part of this project as its already a pretty huge initiative.

As I see it atm, the first step is laying the groundwork in npm for signing packages using some trusted identity, leveraging Sigstore as a vendor-neutral signing service. We'll start by supporting trusted build or workflow identities (CI/CD OIDC id tokens) for new packages.

Once we have this new capability in place we could investigate backfilling this information for legacy packages. For backfilled information the identity would need to be provided by npm instead of the CI/CD system but in principle it could leverage all the same functionality.

Reply:
At the minimum it would represent a point in time snapshot of the current world. While not ideal, I think that does add value.